File: platforms/android/java/app/msime/client/KeyboardScheme.java
package app.msime.client;

import java.util.Arrays;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Set;

/** Shared-Engine keyboard schemes currently exposed by the Android host. */
public enum KeyboardScheme {
    QUANPIN("quanpin", "quanpin", null, "twenty_six_key", "全拼 26 键", "拼", "26"),
    QUANPIN_NINE_KEY("nine_key", "quanpin", null, "nine_key", "全拼 9 键", "拼", "9"),
    XIAOHE("xiaohe", "shuangpin", "xiaohe", "twenty_six_key", "小鹤双拼", "鹤", "双"),
    ZIRANMA("ziranma", "shuangpin", "ziranma", "twenty_six_key", "自然码双拼", "自", "双"),
    MICROSOFT("microsoft", "shuangpin", "microsoft", "twenty_six_key", "微软双拼", "微", "双"),
    SHOUDAO("shoudao", "shuangpin", "shoudao", "twenty_six_key", "首道双拼", "S", "双"),
    WUBI("wubi", "wubi", null, "twenty_six_key", "86 五笔", "五", "86"),
    JAPANESE_NINE_KEY("japanese_nine_key", "japanese", null, "nine_key", "日语 9 键", "あ", "9"),
    JAPANESE("japanese", "japanese", null, "twenty_six_key", "日语 26 键", "あ", "26"),
    HANDWRITING("handwriting", "quanpin", null, "handwriting", "手写", "写", "手"),
    THOUGHTFUL_REPLY("thoughtful_reply", "quanpin", null, "twenty_six_key", "高情商回复", "聊", "AI");

    /** Complete preference values needed for one compare-and-swap update. */
    public record PreferenceMapping(
        String scheme, String lastChineseScheme, String shuangpinProfile,
        String touchKeyboardLayout) {}

    private final String preferenceId;
    private final String engineScheme;
    private final String shuangpinProfile;
    private final String touchKeyboardLayout;
    private final String title;
    private final String glyph;
    private final String badge;

    KeyboardScheme(String preferenceId, String engineScheme, String shuangpinProfile,
                   String touchKeyboardLayout, String title, String glyph, String badge) {
        this.preferenceId = preferenceId;
        this.engineScheme = engineScheme;
        this.shuangpinProfile = shuangpinProfile;
        this.touchKeyboardLayout = touchKeyboardLayout;
        this.title = title;
        this.glyph = glyph;
        this.badge = badge;
    }

    public String preferenceId() { return preferenceId; }
    public String engineScheme() { return engineScheme; }
    public String shuangpinProfile() { return shuangpinProfile; }
    public String touchKeyboardLayout() { return touchKeyboardLayout; }
    public String title() { return title; }
    public String glyph() { return glyph; }
    public String badge() { return badge; }

    public static KeyboardScheme fromPreferenceId(String value) {
        if (value == null) return null;
        for (KeyboardScheme candidate : values()) {
            if (candidate.preferenceId.equals(value)) return candidate;
        }
        return null;
    }

    /** Resolves preference IDs in the fixed Apple order and ignores unknown duplicates. */
    public static List<KeyboardScheme> enabledFromPreferenceIds(List<String> ids) {
        if (ids == null) return List.of(values());
        Set<String> requested = new LinkedHashSet<>(ids);
        List<KeyboardScheme> enabled = Arrays.stream(values())
            .filter(candidate -> requested.contains(candidate.preferenceId)).toList();
        return enabled.isEmpty() ? List.of(QUANPIN) : enabled;
    }

    /** Shared selected is authoritative; otherwise preserve the applied scheme or use first enabled. */
    public static KeyboardScheme resolveEnabledSelection(
            KeyboardScheme applied, String selectedPreferenceId, List<KeyboardScheme> enabled) {
        List<KeyboardScheme> available = enabled == null || enabled.isEmpty()
            ? List.of(QUANPIN) : enabled;
        KeyboardScheme selected = fromPreferenceId(selectedPreferenceId);
        if (selected != null && available.contains(selected)) return selected;
        if (selectedPreferenceId == null && applied != null && available.contains(applied)) return applied;
        return available.get(0);
    }

    public static KeyboardScheme fromHostSelection(String value, boolean thoughtfulEnabled,
                                                    KeyboardScheme engineSelection) {
        if (thoughtfulEnabled && THOUGHTFUL_REPLY.name().equals(value)
                && engineSelection == QUANPIN) return THOUGHTFUL_REPLY;
        return engineSelection;
    }

    public static KeyboardScheme fromPreferences(String scheme, String profile, String touchLayout) {
        if ("quanpin".equals(scheme) && "handwriting".equals(touchLayout)) return HANDWRITING;
        if ("quanpin".equals(scheme) && "nine_key".equals(touchLayout)) return QUANPIN_NINE_KEY;
        if ("japanese".equals(scheme) && "nine_key".equals(touchLayout)) return JAPANESE_NINE_KEY;
        if ("shuangpin".equals(scheme)) {
            for (KeyboardScheme candidate : values()) {
                if (profile != null && profile.equals(candidate.shuangpinProfile)) return candidate;
            }
            return XIAOHE;
        }
        for (KeyboardScheme candidate : values()) {
            if (candidate != THOUGHTFUL_REPLY && candidate.shuangpinProfile == null && candidate.engineScheme.equals(scheme)
                    && !"nine_key".equals(candidate.touchKeyboardLayout)) return candidate;
        }
        return QUANPIN;
    }

    public PreferenceMapping mapping(String currentLastChineseScheme, String currentProfile) {
        String profile = normalizedProfile(currentProfile);
        if (shuangpinProfile != null) profile = shuangpinProfile;
        String lastChinese = isChineseScheme(currentLastChineseScheme)
            ? currentLastChineseScheme : "quanpin";
        if (!"japanese".equals(engineScheme)) lastChinese = engineScheme;
        return new PreferenceMapping(engineScheme, lastChinese, profile, touchKeyboardLayout);
    }

    private static boolean isChineseScheme(String value) {
        return "quanpin".equals(value) || "shuangpin".equals(value) || "wubi".equals(value);
    }

    private static String normalizedProfile(String value) {
        if ("ziranma".equals(value) || "microsoft".equals(value) || "shoudao".equals(value)) {
            return value;
        }
        return "xiaohe";
    }
}

File: platforms/harmony/entry/src/main/ets/keyboard/KeyboardScheme.ts
/**
 * Keyboard schemes the host exposes, ported from
 * platforms/android/java/app/msime/client/KeyboardScheme.java.
 *
 * Java spells this as an enum carrying fields. ArkTS enums hold only a value, so each scheme is a
 * frozen record and SCHEMES preserves the declaration order the Apple hosts also rely on.
 */
export interface SchemeDefinition {
  readonly id: string;
  readonly preferenceId: string;
  readonly engineScheme: string;
  readonly shuangpinProfile: string | null;
  readonly touchKeyboardLayout: string;
  readonly title: string;
  readonly glyph: string;
  readonly badge: string;
}

/** Complete preference values needed for one compare-and-swap update. */
export interface PreferenceMapping {
  readonly scheme: string;
  readonly lastChineseScheme: string;
  readonly shuangpinProfile: string;
  readonly touchKeyboardLayout: string;
}

const QUANPIN: SchemeDefinition = {
  id: "QUANPIN",
  preferenceId: "quanpin",
  engineScheme: "quanpin",
  shuangpinProfile: null,
  touchKeyboardLayout: "twenty_six_key",
  title: "全拼 26 键",
  glyph: "拼",
  badge: "26",
};
const QUANPIN_NINE_KEY: SchemeDefinition = {
  id: "QUANPIN_NINE_KEY",
  preferenceId: "nine_key",
  engineScheme: "quanpin",
  shuangpinProfile: null,
  touchKeyboardLayout: "nine_key",
  title: "全拼 9 键",
  glyph: "拼",
  badge: "9",
};
const XIAOHE: SchemeDefinition = {
  id: "XIAOHE",
  preferenceId: "xiaohe",
  engineScheme: "shuangpin",
  shuangpinProfile: "xiaohe",
  touchKeyboardLayout: "twenty_six_key",
  title: "小鹤双拼",
  glyph: "鹤",
  badge: "双",
};
const ZIRANMA: SchemeDefinition = {
  id: "ZIRANMA",
  preferenceId: "ziranma",
  engineScheme: "shuangpin",
  shuangpinProfile: "ziranma",
  touchKeyboardLayout: "twenty_six_key",
  title: "自然码双拼",
  glyph: "自",
  badge: "双",
};
const MICROSOFT: SchemeDefinition = {
  id: "MICROSOFT",
  preferenceId: "microsoft",
  engineScheme: "shuangpin",
  shuangpinProfile: "microsoft",
  touchKeyboardLayout: "twenty_six_key",
  title: "微软双拼",
  glyph: "微",
  badge: "双",
};
const SHOUDAO: SchemeDefinition = {
  id: "SHOUDAO",
  preferenceId: "shoudao",
  engineScheme: "shuangpin",
  shuangpinProfile: "shoudao",
  touchKeyboardLayout: "twenty_six_key",
  title: "首道双拼",
  glyph: "S",
  badge: "双",
};
const WUBI: SchemeDefinition = {
  id: "WUBI",
  preferenceId: "wubi",
  engineScheme: "wubi",
  shuangpinProfile: null,
  touchKeyboardLayout: "twenty_six_key",
  title: "86 五笔",
  glyph: "五",
  badge: "86",
};
const JAPANESE_NINE_KEY: SchemeDefinition = {
  id: "JAPANESE_NINE_KEY",
  preferenceId: "japanese_nine_key",
  engineScheme: "japanese",
  shuangpinProfile: null,
  touchKeyboardLayout: "nine_key",
  title: "日语 9 键",
  glyph: "あ",
  badge: "9",
};
const JAPANESE: SchemeDefinition = {
  id: "JAPANESE",
  preferenceId: "japanese",
  engineScheme: "japanese",
  shuangpinProfile: null,
  touchKeyboardLayout: "twenty_six_key",
  title: "日语 26 键",
  glyph: "あ",
  badge: "26",
};
const HANDWRITING: SchemeDefinition = {
  id: "HANDWRITING",
  preferenceId: "handwriting",
  engineScheme: "quanpin",
  shuangpinProfile: null,
  touchKeyboardLayout: "handwriting",
  title: "手写",
  glyph: "写",
  badge: "手",
};
const THOUGHTFUL_REPLY: SchemeDefinition = {
  id: "THOUGHTFUL_REPLY",
  preferenceId: "thoughtful_reply",
  engineScheme: "quanpin",
  shuangpinProfile: null,
  touchKeyboardLayout: "twenty_six_key",
  title: "高情商回复",
  glyph: "聊",
  badge: "AI",
};

export class KeyboardScheme {
  static readonly QUANPIN: SchemeDefinition = QUANPIN;
  static readonly QUANPIN_NINE_KEY: SchemeDefinition = QUANPIN_NINE_KEY;
  static readonly XIAOHE: SchemeDefinition = XIAOHE;
  static readonly ZIRANMA: SchemeDefinition = ZIRANMA;
  static readonly MICROSOFT: SchemeDefinition = MICROSOFT;
  static readonly SHOUDAO: SchemeDefinition = SHOUDAO;
  static readonly WUBI: SchemeDefinition = WUBI;
  static readonly JAPANESE_NINE_KEY: SchemeDefinition = JAPANESE_NINE_KEY;
  static readonly JAPANESE: SchemeDefinition = JAPANESE;
  static readonly HANDWRITING: SchemeDefinition = HANDWRITING;
  static readonly THOUGHTFUL_REPLY: SchemeDefinition = THOUGHTFUL_REPLY;

  /** Declaration order is the fixed order the pickers render. */
  static readonly SCHEMES: SchemeDefinition[] = [
    QUANPIN,
    QUANPIN_NINE_KEY,
    XIAOHE,
    ZIRANMA,
    MICROSOFT,
    SHOUDAO,
    WUBI,
    JAPANESE_NINE_KEY,
    JAPANESE,
    HANDWRITING,
    THOUGHTFUL_REPLY,
  ];

  static fromPreferenceId(value: string | null): SchemeDefinition | null {
    if (value === null) {
      return null;
    }
    for (const candidate of KeyboardScheme.SCHEMES) {
      if (candidate.preferenceId === value) {
        return candidate;
      }
    }
    return null;
  }

  /** Resolves preference IDs in the fixed order and ignores unknown duplicates. */
  static enabledFromPreferenceIds(ids: string[] | null): SchemeDefinition[] {
    if (ids === null) {
      return KeyboardScheme.SCHEMES;
    }
    const enabled: SchemeDefinition[] = [];
    for (const candidate of KeyboardScheme.SCHEMES) {
      if (ids.includes(candidate.preferenceId)) {
        enabled.push(candidate);
      }
    }
    return enabled.length === 0 ? [QUANPIN] : enabled;
  }

  /** Shared selection is authoritative; otherwise preserve the applied scheme or use first enabled. */
  static resolveEnabledSelection(
    applied: SchemeDefinition | null,
    selectedPreferenceId: string | null,
    enabled: SchemeDefinition[] | null,
  ): SchemeDefinition {
    const available: SchemeDefinition[] =
      enabled === null || enabled.length === 0 ? [QUANPIN] : enabled;
    const selected: SchemeDefinition | null = KeyboardScheme.fromPreferenceId(selectedPreferenceId);
    if (selected !== null && available.includes(selected)) {
      return selected;
    }
    if (selectedPreferenceId === null && applied !== null && available.includes(applied)) {
      return applied;
    }
    return available[0];
  }

  /** The Engine preference mapping needed when the shared picker changed the fallback. */
  static mappingForRuntimeSelection(
    applied: SchemeDefinition | null,
    selected: SchemeDefinition | null,
    currentLastChineseScheme: string | null,
    currentProfile: string | null,
  ): PreferenceMapping | null {
    if (selected === null || selected === THOUGHTFUL_REPLY || selected === applied) {
      return null;
    }
    return KeyboardScheme.mapping(selected, currentLastChineseScheme, currentProfile);
  }

  static fromHostSelection(
    value: string | null,
    thoughtfulEnabled: boolean,
    engineSelection: SchemeDefinition,
  ): SchemeDefinition {
    if (thoughtfulEnabled && value === THOUGHTFUL_REPLY.id && engineSelection === QUANPIN) {
      return THOUGHTFUL_REPLY;
    }
    return engineSelection;
  }

  static fromPreferences(
    scheme: string,
    profile: string | null,
    touchLayout: string,
  ): SchemeDefinition {
    if (scheme === "quanpin" && touchLayout === "handwriting") {
      return HANDWRITING;
    }
    if (scheme === "quanpin" && touchLayout === "nine_key") {
      return QUANPIN_NINE_KEY;
    }
    if (scheme === "japanese" && touchLayout === "nine_key") {
      return JAPANESE_NINE_KEY;
    }
    if (scheme === "shuangpin") {
      for (const candidate of KeyboardScheme.SCHEMES) {
        if (profile !== null && profile === candidate.shuangpinProfile) {
          return candidate;
        }
      }
      return XIAOHE;
    }
    for (const candidate of KeyboardScheme.SCHEMES) {
      if (
        candidate !== THOUGHTFUL_REPLY &&
        candidate.shuangpinProfile === null &&
        candidate.engineScheme === scheme &&
        candidate.touchKeyboardLayout !== "nine_key"
      ) {
        return candidate;
      }
    }
    return QUANPIN;
  }

  static mapping(
    scheme: SchemeDefinition,
    currentLastChineseScheme: string | null,
    currentProfile: string | null,
  ): PreferenceMapping {
    let profile: string = KeyboardScheme.normalizedProfile(currentProfile);
    if (scheme.shuangpinProfile !== null) {
      profile = scheme.shuangpinProfile;
    }
    let lastChinese: string =
      KeyboardScheme.isChineseScheme(currentLastChineseScheme) && currentLastChineseScheme !== null
        ? currentLastChineseScheme
        : "quanpin";
    if (scheme.engineScheme !== "japanese") {
      lastChinese = scheme.engineScheme;
    }
    return {
      scheme: scheme.engineScheme,
      lastChineseScheme: lastChinese,
      shuangpinProfile: profile,
      touchKeyboardLayout: scheme.touchKeyboardLayout,
    };
  }

  /**
   * The Engine's runtime scheme id as the name the shared policies compare against.
   *
   * The view is handed a number by the Engine view and the policies are written in terms of the
   * scheme names the preference document uses; without this the two silently fail to match, and a
   * policy that refuses something for Japanese never refuses it at all.
   */
  static engineSchemeName(id: number): string {
    switch (id) {
      case 1:
        return "shuangpin";
      case 2:
        return "wubi";
      case 3:
        return "japanese";
      default:
        return "quanpin";
    }
  }

  /** Handwriting is a Chinese typing face, not the English, symbol or local-utility keyboard. */
  static usesHandwritingFace(
    scheme: SchemeDefinition,
    english: boolean,
    symbols: boolean,
    localMode: string,
  ): boolean {
    return (
      scheme.touchKeyboardLayout === "handwriting" && !english && !symbols && localMode === "none"
    );
  }

  /** Local utilities take literal alphabetic keys even when the saved touch layout is nine-key. */
  static usesNineKeyFace(nineKey: boolean, localMode: string): boolean {
    return nineKey && localMode === "none";
  }

  private static isChineseScheme(value: string | null): boolean {
    return value === "quanpin" || value === "shuangpin" || value === "wubi";
  }

  private static normalizedProfile(value: string | null): string {
    if (value === "ziranma" || value === "microsoft" || value === "shoudao") {
      return value;
    }
    return "xiaohe";
  }
}
